### PR TITLE
[WIP] Expose prometheus metrics with gitea

### PR DESCRIPTION
**Pull request by metalmatze**
We already have metrics displayed on the admin dashboard. But I want to be able to create metrics from them. A lot more metrics can be added in the future. I think about things like: pushes & pulls or whatever comes to mind.

[Prometheus](https://prometheus.io/) is the way to go with metrics in 2016 and of course in 2017.
After being added to Kubernetes, Docker and now Gitlab. I want to have metrics in Gitea too.

Metrics from the dashboard below are exposed like the following:
screenshot from 2017-01-15 22-16-17

```
# HELP gitea_accesses Number of Accesses
# TYPE gitea_accesses gauge
gitea_accesses 0
# HELP gitea_actions Number of Actions
# TYPE gitea_actions gauge
gitea_actions 3
# HELP gitea_attachments Number of Attachments
# TYPE gitea_attachments gauge
gitea_attachments 0
# HELP gitea_comments Number of Comments
# TYPE gitea_comments gauge
gitea_comments 0
# HELP gitea_follows Number of Follows
# TYPE gitea_follows gauge
gitea_follows 0
# HELP gitea_hooktasks Number of HookTasks
# TYPE gitea_hooktasks gauge
gitea_hooktasks 0
# HELP gitea_issues Number of Issues
# TYPE gitea_issues gauge
gitea_issues 0
# HELP gitea_labels Number of Labels
# TYPE gitea_labels gauge
gitea_labels 0
# HELP gitea_loginsources Number of LoginSources
# TYPE gitea_loginsources gauge
gitea_loginsources 0
# HELP gitea_milestones Number of Milestones
# TYPE gitea_milestones gauge
gitea_milestones 0
# HELP gitea_mirrors Number of Mirrors
# TYPE gitea_mirrors gauge
gitea_mirrors 0
# HELP gitea_oauths Number of Oauths
# TYPE gitea_oauths gauge
gitea_oauths 0
# HELP gitea_organizations Number of Organizations
# TYPE gitea_organizations gauge
gitea_organizations 1
# HELP gitea_publickeys Number of PublicKeys
# TYPE gitea_publickeys gauge
gitea_publickeys 0
# HELP gitea_releases Number of Releases
# TYPE gitea_releases gauge
gitea_releases 0
# HELP gitea_repositories Number of Repositories
# TYPE gitea_repositories gauge
gitea_repositories 1
# HELP gitea_stars Number of Stars
# TYPE gitea_stars gauge
gitea_stars 0
# HELP gitea_teams Number of Teams
# TYPE gitea_teams gauge
gitea_teams 1
# HELP gitea_updatetasks Number of UpdateTasks
# TYPE gitea_updatetasks gauge
gitea_updatetasks 0
# HELP gitea_users Number of Users
# TYPE gitea_users gauge
gitea_users 1
# HELP gitea_watches Number of Watches
# TYPE gitea_watches gauge
gitea_watches 1
# HELP gitea_webhooks Number of Webhooks
# TYPE gitea_webhooks gauge
gitea_webhooks 0
```

What's missing in WIP:

* Metrics are exposed on `/metrics` and are currently publicly visible. I would be great to generate a token on start and display for admins to grab from the admin web UI. Add a middleware to check the token and only then display the metrics via http. E.g. `/metrics?token=abc123`.

### CODE REVIEW

**Comment by metalmatze** [comment]
Ping @juliusv to review the little prometheus pieces. :blush:

**Comment by juliusv** [comment]
Seems like these metrics are only DB object statistics so far. Maybe make that fact clearer in both the metric names and descriptions? For example, if I read `gitea_accesses`, as an unsuspecting user I might as well expect a counter which just gets incremented any time there is an access. And later on you will probably add non-DB-stat metrics as well and those should be differentiated somehow.

That also raises the question of whether something like `gitea_accesses` should be a gauge or a counter. It's a bit on the edge, as it's nothing we're directly counting ourselves, but on the other hand it should never go down, and it does make sense to take the `rate()` of it, so maybe it makes more sense to have as a counter?

**Comment by juliusv** [comment]
(of course, if you do decide to make it a counter, the `_total` suffix should be added)

**Comment by juliusv** [comment]
Looks good otherwise though.

**Comment by metalmatze** [comment]
Thanks for the very fast response.
You're right about the data just coming from database and we should probably make it more obvious. :+1: 
Waiting for other comments now. :blush:

**Comment by brian-brazil** [comment]
> That also raises the question of whether something like gitea_accesses should be a gauge or a counter. > It's a bit on the edge, as it's nothing we're directly counting ourselves, but on the other hand it should never go down, and it does make sense to take the rate() of it, so maybe it makes more sense to have as a counter?

They're gauges as you care about the exact value, and most of those can go down.

**Comment by tboerger** [comment]
Cool! You get my LGTM because I trust you that this gets awesome ;)
But nevertheless vendor.json got conflicts.

**Comment by bkcsoft** [comment]
Woho 🎉 L-G-T-M once you believe it's not a WIP anymore 😄

File: modules/metrics/statistic.go
Comment (bkcsoft) [inline modules/metrics/statistic.go:12]: Maybe this file should be named "collector.go" instead to reflect the content better? :)

**Review by bkcsoft — CHANGES_REQUESTED** [review]

**Comment by metalmatze** [comment]
So before merging this basic exporter we need to generate a token and display that in the UI. Also check it in a middleware to be the correct one. Otherwise everyones statistics would be public, if gitea is public.
Probably show the token on `/admin/monitor` and add a simple button to regenerate the token, if needed. The token needs to be stored somewhere persistent (the db?).
Any ideas?

**Comment by appleboy** [comment]
I have tested this PR. It is working for me. Thanks @metalmatze LGTM

**Comment by bkcsoft** [comment]
@metalmatze if it's generated once my vote goes for `app.ini` (same as we do for `SECRET` 🙂 )

File: cmd/web.go
Comment (bkcsoft) [inline cmd/web.go:618]: This should probably have `adminReq`... or some other token-based auth

**Review by bkcsoft — CHANGES_REQUESTED** [review]
Security 😄

**Comment by metalmatze** [comment]
@bkcsoft So, if we generate a token and put it into the `app.ini` should we nevertheless display that in the UI somewhere, or simply require the user to pick it from that file?

**Comment by lunny** [comment]
Is this still wip?

**Comment by metalmatze** [comment]
Yes. I don't really have decided on how to handle the generated token.

**Comment by lunny** [comment]
let's move this to v1.2.

**Comment by bkcsoft** [comment]
@metalmatze For now just show it in the admin-panel, maybe obfuscate it with a button that ways `Press here to show monitor-token`. As for regenerating we could wait with that, one could always change it in the config-file. (I'm just making suggestions to speed this up a bit 😄)

**Comment by sapk** [comment]
like @bkcsoft said it could be set in app.ini and display under /admin/config along side the rest of the config. This page is restricted to admin and they should be the one able to use the token.
Looking forward to having this merge. ^^

**Comment by bkcsoft** [comment]
@metalmatze Any new on this? Seems stale and we don't want to give people false hopes 😉

**Pull request closed without merging**
